### ISSUE TITLE
Publish as Home Assistant addon

### ISSUE THREAD

**Issue by the-mentor** [toverainc/willow-inference-server#76]
the title says it all having WIS as home assistant addon will lower the barrier to entry for many people.

This project is amazing keep up the amazing work !!

**Comment by kristiankielhofner** [comment]
Thanks!

We're actually evaluating our general Home Assistant add-on/component approach. We would definitely like to get Willow/WIS more well integrated with HA, we just want to solidify some things on our side first.

**Comment by lundyfpv** [comment]
I'm 100% on board with this. I have my Tesla P4 just waiting to be dropped into my HA server.

**Comment by kristiankielhofner** [comment]
The Home Assistant component has nothing to do with where and how WIS runs. Your Tesla P4 won't be utilized by Home Assistant with or without a Willow Home Assistant component.

The Willow Home Assistant component will (essentially) be a thin network proxy layer that hooks into Home Assistant to provide tighter integration with Home Assistant. It has nothing to do with inference or anything WIS does.

If you have a Tesla P4 it can be utilized by WIS today.

**Comment by lordratner** [comment]
> The Home Assistant component has nothing to do with where and how WIS runs. Your Tesla P4 won't be utilized by Home Assistant with or without a Willow Home Assistant component.
> 
> The Willow Home Assistant component will (essentially) be a thin network proxy layer that hooks into Home Assistant to provide tighter integration with Home Assistant. It has nothing to do with inference or anything WIS does.
> 
> If you have a Tesla P4 it can be utilized by WIS today.

I think what he meant was "add-on" rather than "integration" or component, which in HAOS is a Home Assistant Supervisor-operated container with a service in it. As an example, you can use the Mosquitto MQTT broker addon which starts a container with Mosquitto, but monitored, updated, and accessed through Home Assistant. 

This is probably impractical for WIS due to the resource requirements (and GPU option, which I don't think Home Assistant covers) based on how many people are running HAOS on a Raspberry Pi.

**Comment by lundyfpv** [comment]
Yep addon/container is what we mean. The question of does HA allow Gpu pass through for add-ons is something I do not know the answer to though.

**Comment by lordratner** [comment]
I think direct integration with HA following the Add-on route is a complete non-starter, for the reasons discussed above. 

However HA has been making huge changes with voice this year, so I don't think an add-on is the right way to look at this anyways. They are now using "pipelines" I believe, which is just a fancy way of saying you set up a voice assistant using three components: a conversation agent, STT, and TTS. You can pick from any available agent for each job.

![pipeline](https://github.com/toverainc/willow-inference-server/assets/49489398/a511a743-8e06-4717-8a75-b78df5e6bd27)

In this case, Piper and Whisper are add-ons running on the HAOS hardware, but that is not necessary. It's just the easy-button answer for most of their users, again, running on Raspberry Pis. Whisper has addon configuration settings for what models to use and beam size, with only one model working on a Raspberry Pi.

These Addons then use the ["Wyoming Protocol"](https://github.com/rhasspy/rhasspy3/blob/master/docs/wyoming.md) **integration** to tie the two together. This is where WIS would integrate into HAOS, I think. You just put the IP and Port of the service, and HA makes it available to add to an assistant pipeline. 

The question then becomes, are WIS and the Wyoming Protocol compatible?

**Comment by kristiankielhofner** [comment]
I think we're talking past each other - everything in your screenshot is fundamentally an HA integration component that talks to some external resource:

- Nabu Casa cloud for their Azure text to speech.
- Faster-whisper is a connection to their Whisper container, which again lives separately.
- Same with Piper.
- Same with the OpenAI components for supported functionality.

A WIS HA integration component would expose the capabilities of WIS to HA as these components due today. WIS for speech to text, text to speech, etc would just become another option in the drop downs of the screenshot you provided.

These various add-ons do not solely use the Wyoming Protocol, in fact they overwhelmingly[ do not](https://www.home-assistant.io/integrations/#voice).

WIS and Willow will not be directly implementing the Wyoming Protocol for various reasons:

1) While it is "open" it's also essentially proprietary to the Home Assistant ecosystem. The state of a protocol being "open" comes into significant question when it's not implemented by anything else outside of the ecosystem of the creators.

2) It's fundamentally broken. The entire concept of MQTT for voice transport is problematic for a variety of inherent reasons. The creators of Wyoming would be well served to look at other mature technology areas that do media transport - media streaming, voice/video over IP, etc. Entire industries across use cases have learned many hard lessons from prior failed approaches like Wyoming but the Home Assistant ecosystem seems dedicated to repeating all of these mistakes (and more) from decades ago, while somehow thinking their results are going to be different. There isn't a single implementation outside of the smart home ecosystem using MQTT or anything like it for media transport - and for many good reasons.

3) It's new and unproven. Combined with my position above, I think it will be very difficult (if not impossible) for Wyoming based approaches (as well as others taken within the HA ecosystem) to deliver anything resembling the level of experience Willow (again, at two months old) does today. Let alone the extremely rapid progress we're making on things so far beyond what HA is capable of today they're not even on the radar. There are many significantly better existing protocols and approaches that have matured over decades (like the ones Willow uses).

I've noted before we don't see there being any reason Willow has to be in any kind of conflict with HA, the community, the team, or the ecosystem but I find it curious to repeatedly have these conversations when Willow at two months old (based on my decades of experience in these areas and more) already provides a quality of experience that (when it really comes down to it) is leagues beyond what the rest of the open source ecosystem is doing in this area. There are very good reasons for that, and these fundamental issues and more contribute to (or are the direct cause of) this significant gap in capability and user experience.

**Comment by lordratner** [comment]
> I think we're talking past each other - everything in your screenshot is fundamentally an HA integration component that talks to some external resource:

Yeah definitely, that was my point. In the case of the screenshot Whisper and Piper are running as container addons in HAOS, but WIS would certainly have to run externally. It would just be selected in the pipeline interface I posted  the screenshot of.

No clue about Wyoming, but the info you shared is fascinating. I just noticed that HA was using it. It sounds like WIS will probably just use an integration. Put in the IP, port, and maybe something for authentication (no clue or input, I'm sure you have that figured). Boom. Now it's available for use.

I think part of the reason we're talking past each other is because the original poster was also fairly unclear about add-ons and what capabilities they are limited to.

**Comment by the-mentor** [comment]
Home assistant addons are just a simple way to run a pre-defined container  images for stuff like WIS they also allow to use hardware like skyconnect etc
An addon will make it very simple for users to get started and self host WIS. 

Also if WIS can also use coral ai usb accelerator rather then GPU it will make it even easier.

Just some ideas whatever the project goes with works for me since I know how to run docker I just think it will be great for the ecosystem

Thanks

**Comment by kristiankielhofner** [comment]
@lordratner - Exactly. If you look at the HA components for all of the other voice integrations they essentially say "Point at a server, provide some details, we abstract the rest". The Wyoming situation is unfortunate and I hope it improves (they made changes with V3 but I'd argue it's just as problematic). One of the issues we're having is Willow being extremely new. Rhasspy (as an example) is four years old... When I first posted the EARLY EARLY preview announcement of Willow to Hacker News it spread very far and wide to audiences I didn't expect and we weren't prepared to deal with. As anyone can tell just from looking at the "install" instructions Willow and WIS are not intended for casual HA users and we're reminded of that on a daily basis. That said we're happy to work with the unintended/expected broader user base and we've re-prioritized development priorities to better serve these users - much earlier than expected. My thinking and expectation with the initial release was for a couple of dozen developers to work through the early pains of Willow. That clearly hasn't been the case (for good, bad, and ugly).

@the-mentor Again, on point. In the end all of this stuff (WIS, Piper, etc) are just docker containers. The issue of running WIS on HAOS (as bad of an idea as that may be) really just comes down to documentation. Support for Coral is another recurring question. In short, it's fundamentally impossible (tiny memory, TFLite only) to do broad speech recognition or speech generation on a Coral. The on ESP BOX pre-defined command detection already supports at least 2x (~150 vs 400) as many pre-defined commands vs Coral.

**Comment by tensiondriven** [comment]
+1 for any Home Assistant integration being a lightweight wrapper/API layer available via HACS.  Eventually someone may roll willow into a HA addon, but it doesn't make sense to me to try to support it as an addon.

To enable someone from the community to do this, perhaps providing a good REST API with webhooks and decent documentation is all it would take.  (I'd use such a REST API + Webhook for my own project, actually.)

**Comment by ccgauvin94** [comment]
Seeing what HA is doing now with an on-device Assistant option (to replace Google Assistant on Android, for instance), I can only dream of what it'd be like to have that routing to a Willow Inference Server running Vicuna (or perhaps LLama2). 

This is an incredibly impressive project. My mind is blown. I haven't felt like this about computers and technology in such a long time, I feel like a kid again, imagining all the possibilities of what this could do. Congratulations on the unbelievable release.

**Comment by tensiondriven** [comment]
HA *just* added the ability for "services" to return values in the latest release.  I can only imagine this was done to enable language models to be called with a given set of args and to return something - either a text string, json, function call, etc.

Exciting times.
